### ISSUE TITLE
testing problems

### ISSUE THREAD

**Issue by max0312** [JuliaWolleb/diffusion-anomaly#5]
Hi, thanks for your great job
I have some questions after training and testing on the brats dataset
 
1. Did you normalize the seg part(data/brats/training/000001/brats_train_001_seg_080_w.nii.gz) during preprocessing? I see that the training sample set you gave has not been normalized. Will this affect the cross_entropy calculation during classifier training?
 
2. I found that the results generated by using the brats2update010000.pt and modelbratsclass010000.pt saved at the beginning of the training are better than those of the model saved in the later training. What is the reason for this? (What are the specific model parameters you used in the test?)
 
![image](https://user-images.githubusercontent.com/43068211/203382354-439b9483-8c7f-4df6-9318-8e6668e5b043.png)

3. When testing one by one, the dice coefficient fluctuated greatly, but the auc was always high (0.98~0.99) could you please tell me why?
![image](https://user-images.githubusercontent.com/43068211/203382408-4f4967ca-cd7d-40d0-b808-faa790288374.png)

**Comment by JuliaWolleb** [comment]
Hi
1. The ground truth segmentation is not used during the training. This is only used to compute the dice score during evalutation. All class labels >0 get the class label 1 as "anomalous", as indicated in the line 34 of the file _scripts/evaluation_metrics.py_. All MR slices of the BRATS dataset are normalized to values between 0 and 1.
2. There seems to be a mistake in the example image you gave for _brats2update050000.pt_ and _modelbratsclass070000.pt_. Do you compare the right input vs. the right output?
3. I would check in what cases the dice score was so bad. were these the images where the tumor was really small?

**Comment by Tungthanhlee** [comment]
Hi, Thank you for publishing your work. Im trying to reproduce your work for a week now and couldn't figure out what's wrong with my reproduction. After I trained the classifier and diffusion model, I run the classifier_sample_known.py with checkpoint brats2update050000.pt and modelbratsclass020000.pt. However, I got the results as below. The sampled_output is just the black and white image. I didn't change anything from your code. Could you figure what I'm doing wrong here? Thank you. 
![image](https://user-images.githubusercontent.com/32807741/204393922-b9866638-212c-4a74-9c1a-5ce950058cf9.png)
![image](https://user-images.githubusercontent.com/32807741/204394030-6be6658c-c651-4b16-9b5f-8069cba11e62.png)

**Comment by JuliaWolleb** [comment]
can you run the script classifier_sample_known.py with a classifier_scale of 0? then the output should show the same image as the input.

**Comment by Tungthanhlee** [comment]
I still got output as back and white image. What do you think of in this situation? The input is clearly correct, but the sample output is wrong.

**Comment by Tungthanhlee** [comment]
Do you mind uploading your checkpoints?

**Comment by JuliaWolleb** [comment]
https://www.dropbox.com/sh/up4sqmeqrlwrxkw/AACqgnSCR8p20r6aO_3_Jy94a?dl=0

**Comment by max0312** [comment]
Thanks for your advice  and I rechecked the results
2. I found that the results generated by using the beginning of the training are still better than those of the model saved in the later training.
3. After checking the low dice score, it was found that the original mask was very small.
But I want to know why the auc is still so high when the dice score calculation is low?


![image](https://user-images.githubusercontent.com/43068211/204568956-47f40148-d09f-4544-b91d-295a3d94b9ff.png)
![image](https://user-images.githubusercontent.com/43068211/204569026-e1171084-bfed-4a00-aad1-5bbda1910308.png)

**Comment by Linly1080** [comment]
Thank you very much for your sharing but I encountered a similar problem. 
I used the checkpoints in your link above, but still output black and white images. And there is no obvious noise on reversesample image. For the code, I just modified bratsloader.py. 
Can you give me some advice? 
![image](https://user-images.githubusercontent.com/84452318/226258048-ff9fa976-afd6-4dfc-93dd-7da6d2dd355f.png)

**Comment by Linly1080** [comment]
> I still got output as back and white image. What do you think of in this situation? The input is clearly correct, but the sample output is wrong.

hello, have you solved this problem after using the checkpoints uploaded?

**Comment by JuliaWolleb** [comment]
There is something going wrong with your noise encoding and decoding. How large did you choose L? The image "reversesample" should show a noisy image of noise level L. If you put s=0, the final images should show the same images as the input. Could you check four function "ddim_sample_loop_known"?

**Comment by Linly1080** [comment]
--dataset brats --classifier_scale 100 --noise_level 500 --image_size 256 --num_channels 128 --class_cond True --num_res_blocks 2 --num_heads 1 --learn_sigma True --use_scale_shift_norm False --attention_resolutions 16 --diffusion_steps 1000 --noise_schedule linear --rescale_learned_sigmas False --rescale_timesteps False --image_size 256 --classifier_attention_resolutions 32,16,8 --classifier_depth 4 --classifier_width 32 --classifier_pool attention --classifier_resblock_updown True --classifier_use_scale_shift_norm True --batch_size 1 --num_samples 1 --timestep_respacing ddim1000 --use_ddim True
This is my test set. 
I feel like my denoising and denoising process is reversed. I run classifer_sample_known.py with "--dataset brats --classifier_scale 100 --noise_level 1000 --image_size 256 --num_channels 128 --class_cond True --num_res_blocks 2 --num_heads 1 --learn_sigma True --use_scale_shift_norm False --attention_resolutions 16 --diffusion_steps 1000 --noise_schedule linear --rescale_learned_sigmas False --rescale_timesteps False --image_size 256 --classifier_attention_resolutions 32,16,8 --classifier_depth 4 --classifier_width 32 --classifier_pool attention --classifier_resblock_updown True --classifier_use_scale_shift_norm True". The result is as follow: 
![image](https://user-images.githubusercontent.com/84452318/226296289-95e7e2fc-e8d1-4e75-b0f3-36e6402a68ea.png)
it seems that it add noise when use p_sample()
![image](https://user-images.githubusercontent.com/84452318/226296846-7511f8ad-f076-4680-a86f-5ae83823bd3b.png)

**Comment by Linly1080** [comment]
> There is something going wrong with your noise encoding and decoding. How large did you choose L? The image "reversesample" should show a noisy image of noise level L. If you put s=0, the final images should show the same images as the input. Could you check four function "ddim_sample_loop_known"?

Hello, the function "ddim_sample_loop_known" has not changed. Can you give me more advice?
